


## Weather Analysis Project
**(EDA,Visualizations with Power bi,Conclusion)**

### The Weather Dataset

### Description:
The Weather Dataset is a time-series dataset with per-hour information about the weather condition at a particular location. It records Temperature, dew Point temperature, Relative humidity, Wind speed, Visibility, Pressure and conditions.


This Dataset available as a csv file. We are going to analyze this dataset using Pandas dataframe.

### Workflow:
* Import library
* Analyzing the data
* solve few questions of Data Analytics


In [42]:
# import library
import pandas as pd

In [43]:
# Load dataset
df=pd.read_csv("/content/Weather Data.csv")

In [44]:
df

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa               Weather  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

### Analyzing the data

**.head()**

Its show the first N rows in the data( by default N=5)

In [45]:
df.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

**.shape**

its shows totak number of rows and columns of the dataframe.

In [46]:
df.shape

(8784, 8)

**.index**

This attribute provide the index of the dataframe

In [47]:
df.index

RangeIndex(start=0, stop=8784, step=1)

**columns**

Its shows the name of each columns.

In [48]:
df.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

**.dtypes**

Its shows datatypes of each column.

In [49]:
df.dtypes

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

**.unique()**

In a column,its shows alll the unique values. It can be applied on a sinngle column only.Not on the whole dataframe.

In [50]:
df['Weather'].unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

**.nunique()**

Its shows the total number of unique values in each column.It can be applied on a single column as on whole dataframe.

In [51]:
df.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

**.count**

It show the total no. of non-null values in each column.It can be applied on a single column as well as whole dataframe.

In [52]:
df.count()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

**.value_counts()**

In a column ,it shows all the unique values with their count. It can be applied on single column only.

In [53]:
df['Weather'].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

**.info()**

It provides basic information about the datframe.

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


### Q.1 Find all the unique 'Wind speed' values in the data.

In [55]:
df.head()


Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

In [56]:
df.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

In [57]:
df["Wind Speed_km/h"].nunique()

34

In [58]:
df["Wind Speed_km/h"].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2])

- THis is the unique wind speeds (km/h) in dataset.

### Q.2 Find the number of time when the 'Weather is exactly clear'.

In [59]:
df.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

In [60]:
#1: value_counts()
df.Weather.value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

In [61]:
#2. Filtering

df[df.Weather== 'Clear']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

In [62]:
#3. groupby()

#df.head()
df.groupby('Weather').get_group('Clear')

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

### Q.3 Find the number of time when the 'Weather is exactly 4km/h'.

In [63]:
df[df['Wind Speed_km/h']==4]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

In [64]:
df[df['Wind Speed_km/h']==4].count()

Date/Time           474
Temp_C              474
Dew Point Temp_C    474
Rel Hum_%           474
Wind Speed_km/h     474
Visibility_km       474
Press_kPa           474
Weather             474
dtype: int64

- 474 times the Weather is exactly 4km/h.

### Q.4 Find out all the null values in the data.

In [65]:
df.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

In [66]:
df.notnull().sum()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

- NO null values are present in this dataset.

### Q.5 Rename the column name 'Weather ' of the dataframe to 'Weather condition'.

In [67]:
df.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

In [68]:
df.rename(columns={'Weather':'Weather Condition'}) # onlly for this comands

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa     Weather Condition  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

In [69]:
df.rename(columns={'Weather':'Weather Condition'},inplace=True) # change column name permantly

### Q.6 What is the mean 'visibility'?.

In [70]:
df.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather Condition  
0            8.0     101.24               Fog  
1            8.0     101.24               Fog

In [71]:
df.Visibility_km.mean()

27.664446721311478

- The mean 'visibility' is 27.664446721311478.

### Q.7 What is the standard deviation of 'Pressure ' in this data?


In [72]:
#df.head()
df.Press_kPa.std() # nospace so using '.'

0.8440047459486483

 - The standard deviation of 'Pressure ' in this data is 0.8440047459486483.

### Q.8 What is the variance of 'Relative Humidity ' in this data?

In [73]:
df['Rel Hum_%'].var() #space between name so using []

286.24855019850196

The variance of 'Relative Humidity ' in this data is 286.24855019850196.

### Q.9 Find all instances when 'Snow' was recorded.

In [74]:
#1:value_counts()
#df.head()
df['Weather Condition'].value_counts()

Weather Condition
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle      

In [75]:
#2: Filtering
df[df['Weather Condition']=='Snow']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
55       1/3/2012 7:00   -14.0             -19.5         63               19   
84      1/4/2012 12:00   -13.7             -21.7         51               11   
86      1/4/2012 14:00   -11.3             -19.0         53                7   
87      1/4/2012 15:00   -10.2             -16.3         61               11   
88      1/4/2012 16:00    -9.4             -15.5         61               13   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather Condition  
55             25.0     100.95              Snow  
84             24.1     101.25              Snow  
86             19.3     100.97              Snow  
87              9.7     100.89              Snow  
88             19.3     100.79              Snow  
...             ...        ...               ...  
8779            9.7     100.13              Snow  
8780            9.7     100.03              Snow  
8781            4.8      99.95              Snow  
8782            9.7      99.91              Snow  
8783           11.3      99.89              Snow  

[390 rows x 8 columns]

In [76]:
#3: str.contains
df[df['Weather Condition'].str.contains('Snow')]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
41      1/2/2012 17:00    -2.1              -9.5         57               22   
44      1/2/2012 20:00    -5.6             -13.4         54               24   
45      1/2/2012 21:00    -5.8             -12.8         58               26   
47      1/2/2012 23:00    -7.4             -14.1         59               17   
48       1/3/2012 0:00    -9.0             -16.0         57               28   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather Condition  
41             25.0      99.66      Snow Showers  
44             25.0     100.07      Snow Showers  
45             25.0     100.15      Snow Showers  
47             19.3     100.27      Snow Showers  
48             25.0     100.35      Snow Showers  
...             ...        ...               ...  
8779            9.7     100.13              Snow  
8780            9.7     100.03              Snow  
8781            4.8      99.95              Snow  
8782            9.7      99.91              Snow  
8783           11.3      99.89              Snow  

[583 rows x 8 columns]

In [77]:
df[df['Weather Condition'].str.contains('Snow')].head(50)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
41   1/2/2012 17:00    -2.1              -9.5         57               22   
44   1/2/2012 20:00    -5.6             -13.4         54               24   
45   1/2/2012 21:00    -5.8             -12.8         58               26   
47   1/2/2012 23:00    -7.4             -14.1         59               17   
48    1/3/2012 0:00    -9.0             -16.0         57               28   
50    1/3/2012 2:00   -10.5             -15.8         65               22   
51    1/3/2012 3:00   -11.3             -18.7         54               33   
53    1/3/2012 5:00   -12.9             -19.1         60               22   
54    1/3/2012 6:00   -13.3             -19.3         61               19   
55    1/3/2012 7:00   -14.0             -19.5         63               19   
84   1/4/2012 12:00   -13.7             -21.7         51               11   
86   1/4/2012 14:00   -11.3             -19.0         53                7   
87   1/4/2012 15:00   -10.2             -16.3         61               11   
88   1/4/2012 16:00    -9.4             -15.5         61               13   
89   1/4/2012 17:00    -8.9             -13.2         71                9   
90   1/4/2012 18:00    -8.9             -12.6         75               11   
91   1/4/2012 19:00    -8.4             -12.7         71                9   
92   1/4/2012 20:00    -7.8             -12.1         71                9   
93   1/4/2012 21:00    -7.6             -11.6         73                7   
94   1/4/2012 22:00    -9.5             -12.7         77                6   
95   1/4/2012 23:00    -9.6             -12.6         79                6   
96    1/5/2012 0:00    -8.8             -11.7         79                4   
97    1/5/2012 1:00    -7.5             -10.2         81                0   
98    1/5/2012 2:00    -5.4              -8.3         80                9   
99    1/5/2012 3:00    -5.0              -7.7         81               11   
100   1/5/2012 4:00    -4.1              -6.5         83                9   
101   1/5/2012 5:00    -7.0              -9.5         82                4   
102   1/5/2012 6:00    -8.2             -10.7         82                6   
103   1/5/2012 7:00    -7.1              -9.7         82                9   
104   1/5/2012 8:00    -6.1              -9.1         79               11   
123   1/6/2012 3:00   -10.6             -16.0         64                0   
124   1/6/2012 4:00   -11.3             -16.1         68               15   
125   1/6/2012 5:00   -11.8             -16.0         71               19   
126   1/6/2012 6:00   -12.0             -16.2         71               22   
127   1/6/2012 7:00   -14.4             -16.3         85               22   
128   1/6/2012 8:00   -12.3             -16.2         73               24   
129   1/6/2012 9:00   -12.5             -16.7         71               26   
130  1/6/2012 10:00   -12.3             -16.3         72               28   
131  1/6/2012 11:00   -12.0             -16.0         72               17   
132  1/6/2012 12:00   -11.7             -15.4         74               28   
133  1/6/2012 13:00   -11.9             -15.6         74               20   
134  1/6/2012 14:00   -11.2             -14.8         75               19   
135  1/6/2012 15:00   -11.5             -14.4         79               19   
136  1/6/2012 16:00   -11.6             -14.7         78               19   
137  1/6/2012 17:00   -11.2             -14.3         78               17   
158  1/7/2012 14:00    -4.4              -6.9         83                6   
159  1/7/2012 15:00    -3.7              -6.1         83               11   
160  1/7/2012 16:00    -3.6              -5.9         84                0   
161  1/7/2012 17:00    -3.1              -5.4         84               13   
162  1/7/2012 18:00    -3.2              -5.3         85                6   

     Visibility_km  Press_kPa Weather Condition  
41            25.0      99.66   

In [78]:
df[df['Weather Condition'].str.contains('Snow')].tail(50)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
8680  12/27/2012 16:00    -4.5              -6.2         88               37   
8681  12/27/2012 17:00    -4.2              -5.9         88               32   
8682  12/27/2012 18:00    -4.0              -5.7         88               28   
8683  12/27/2012 19:00    -3.9              -5.6         88               26   
8684  12/27/2012 20:00    -3.7              -5.3         89               37   
8685  12/27/2012 21:00    -3.7              -4.8         92               24   
8686  12/27/2012 22:00    -3.8              -4.6         94               20   
8687  12/27/2012 23:00    -4.0              -5.6         89               24   
8688   12/28/2012 0:00    -4.2              -5.7         89               19   
8689   12/28/2012 1:00    -4.4              -6.6         85               15   
8690   12/28/2012 2:00    -4.3              -6.3         86               11   
8691   12/28/2012 3:00    -4.6              -5.9         91               13   
8692   12/28/2012 4:00    -4.9              -5.9         93                9   
8723  12/29/2012 11:00   -10.9             -12.2         90                7   
8724  12/29/2012 12:00   -10.5             -11.6         92               11   
8725  12/29/2012 13:00   -10.0             -11.1         92               22   
8726  12/29/2012 14:00    -9.3             -10.5         91               22   
8727  12/29/2012 15:00    -8.8             -10.0         91               20   
8728  12/29/2012 16:00    -8.5              -9.9         90               24   
8729  12/29/2012 17:00    -9.0             -10.4         90               19   
8730  12/29/2012 18:00    -9.3             -10.9         88               26   
8731  12/29/2012 19:00    -9.5             -11.2         87               26   
8732  12/29/2012 20:00    -9.7             -11.6         86               24   
8733  12/29/2012 21:00    -9.8             -11.8         85               24   
8734  12/29/2012 22:00   -10.1             -11.6         89               15   
8735  12/29/2012 23:00   -10.0             -12.0         85               20   
8736   12/30/2012 0:00    -9.6             -11.3         87               13   
8737   12/30/2012 1:00    -9.4             -10.5         92                9   
8738   12/30/2012 2:00    -9.3             -10.4         92                9   
8739   12/30/2012 3:00    -9.1             -10.4         90               11   
8740   12/30/2012 4:00    -9.3             -10.6         90               13   
8741   12/30/2012 5:00    -9.1             -10.4         90               11   
8742   12/30/2012 6:00    -9.3             -10.8         89               17   
8767   12/31/2012 7:00    -9.3             -11.3         85                0   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8771  12/31/2012 11:00    -6.7              -7.9         91                9   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   
8774  12/31/2012 14:00    -3.4              -5.7         84                6   
8775  12/31/2012 15:00    -2.3              -4.6         84                9   
8776  12/31/2012 16:00    -1.4              -4.0         82               13   
8777  12/31/2012 17:00    -1.1              -3.3         85               19   
8778  12/31/2012 18:00    -1.3              -3.1         88               17   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2

### Q.10 Find all instances when 'Wind speed is above 24' and 'Visibility'is 25 .

In [79]:
df[(df['Wind Speed_km/h'] > 24) & (df['Visibility_km']==25)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
23      1/1/2012 23:00     5.3               2.0         79               30   
24       1/2/2012 0:00     5.2               1.5         77               35   
25       1/2/2012 1:00     4.6               0.0         72               39   
26       1/2/2012 2:00     3.9              -0.9         71               32   
27       1/2/2012 3:00     3.7              -1.5         69               33   
...                ...     ...               ...        ...              ...   
8705  12/28/2012 17:00    -8.6             -12.0         76               26   
8753  12/30/2012 17:00   -12.1             -15.8         74               28   
8755  12/30/2012 19:00   -13.4             -16.5         77               26   
8759  12/30/2012 23:00   -12.1             -15.1         78               28   
8760   12/31/2012 0:00   -11.1             -14.4         77               26   

      Visibility_km  Press_kPa Weather Condition  
23             25.0      99.31            Cloudy  
24             25.0      99.26      Rain Showers  
25             25.0      99.26            Cloudy  
26             25.0      99.26     Mostly Cloudy  
27             25.0      99.30     Mostly Cloudy  
...             ...        ...               ...  
8705           25.0     101.34      Mainly Clear  
8753           25.0     101.26      Mainly Clear  
8755           25.0     101.47      Mainly Clear  
8759           25.0     101.52     Mostly Cloudy  
8760           25.0     101.51            Cloudy  

[308 rows x 8 columns]

### Q.11 What is minimum and maximum value of each column against each 'Weather condition'

In [80]:
df.groupby('Weather Condition').min()

Date/Time  Temp_C  \
Weather Condition                                                   
Clear                                      1/11/2012 1:00   -23.3   
Cloudy                                     1/1/2012 17:00   -21.4   
Drizzle                                   1/23/2012 21:00     1.1   
Drizzle,Fog                               1/23/2012 20:00     0.0   
Drizzle,Ice Pellets,Fog                   12/17/2012 9:00     0.4   
Drizzle,Snow                             12/17/2012 15:00     0.9   
Drizzle,Snow,Fog                         12/18/2012 21:00     0.3   
Fog                                         1/1/2012 0:00   -16.0   
Freezing Drizzle                          1/13/2012 10:00    -9.0   
Freezing Drizzle,Fog                        1/1/2012 2:00    -6.4   
Freezing Drizzle,Haze                      2/1/2012 11:00    -5.8   
Freezing Drizzle,Snow                      1/13/2012 3:00    -8.3   
Freezing Fog                               1/22/2012 6:00   -19.0   
Freezing Rain                             1/13/2012 11:00    -6.5   
Freezing Rain,Fog                         1/17/2012 23:00    -6.1   
Freezing Rain,Haze                         2/1/2012 14:00    -4.9   
Freezing Rain,Ice Pellets,Fog             12/17/2012 3:00    -2.6   
Freezing Rain,Snow Grains                  1/13/2012 9:00    -5.0   
Haze                                      1/22/2012 12:00   -11.5   
Mainly Clear                              1/10/2012 11:00   -22.8   
Moderate Rain,Fog                         12/10/2012 8:00     1.7   
Moderate Snow                             1/12/2012 15:00    -6.3   
Moderate Snow,Blowing Snow               12/27/2012 10:00    -5.5   
Mostly Cloudy                              1/1/2012 16:00   -23.2   
Rain                                       1/1/2012 18:00     0.3   
Rain Showers                               1/1/2012 22:00     1.6   
Rain Showers,Fog                          10/20/2012 3:00    12.8   
Rain Showers,Snow Showers                  11/4/2012 8:00     2.1   
Rain,Fog                                  1/23/2012 18:00     0.0   
Rain,Haze                                  3/13/2012 7:00     4.0   
Rain,Ice Pellets                          12/18/2012 5:00     0.6   
Rain,Snow                                  1/10/2012 5:00     0.6   
Rain,Snow Grains                          12/21/2012 0:00     1.9   
Rain,Snow,Fog                             12/8/2012 21:00     0.8   
Rain,Snow,Ice Pellets                     12/21/2012 1:00     0.9   
Snow                                       1/10/2012 1:00   -16.7   
Snow Pellets                             11/24/2012 15:00     0.7   
Snow Showers                               1/12/2012 7:00   -13.3   
Snow Showers,Fog                          12/26/2012 9:00   -11.3   
Snow,Blowing Snow                         1/13/2012 21:00   -12.0   
Snow,Fog                                 12/16/2012 15:00   -10.1   
Snow,Haze                                  2/1/2012 17:00    -4.3   
Snow,Ice Pellets                          12/10/2012 3:00    -4.3   
Thunderstorms                              7/16/2012 1:00    21.6   
Thunderstorms,Heavy Rain Showers           5/29/2012 6:00    10.9   
Thunderstorms,Moderate Rain Showers,Fog    7/17/2012 6:00    19.6   
Thunderstorms,Rain                        5/25/2012 20:00    19.4   
Thunderstorms,Rain Showers                5/29/2012 16:00    11.0   
Thunderstorms,Rain Showers,Fog             6/29/2012 3:00    19.5   
Thunderstorms,Rain,Fog                     7/17/2012 5:00    20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                      
Clear                                               -28.5         20   
Cloudy                                              -26.8         18   
Drizzle                                              -0.2         74   
Drizzle,Fog                                          -1.6         85   
Drizzle,Ice Pellets,Fog    

In [81]:
df.groupby('Weather Condition').max()

Date/Time  Temp_C  \
Weather Condition                                                   
Clear                                       9/9/2012 5:00    32.8   
Cloudy                                     9/9/2012 23:00    30.5   
Drizzle                                    9/30/2012 3:00    18.8   
Drizzle,Fog                                9/30/2012 2:00    19.9   
Drizzle,Ice Pellets,Fog                   12/17/2012 9:00     0.4   
Drizzle,Snow                             12/19/2012 18:00     1.2   
Drizzle,Snow,Fog                          12/22/2012 3:00     1.1   
Fog                                        9/22/2012 0:00    20.8   
Freezing Drizzle                            2/1/2012 5:00    -2.3   
Freezing Drizzle,Fog                      12/10/2012 5:00    -0.3   
Freezing Drizzle,Haze                      2/1/2012 13:00    -5.0   
Freezing Drizzle,Snow                      3/2/2012 12:00    -3.3   
Freezing Fog                               3/17/2012 6:00    -0.1   
Freezing Rain                               2/1/2012 7:00     0.3   
Freezing Rain,Fog                         12/17/2012 1:00     0.1   
Freezing Rain,Haze                         2/1/2012 15:00    -4.9   
Freezing Rain,Ice Pellets,Fog             12/17/2012 3:00    -2.6   
Freezing Rain,Snow Grains                  1/13/2012 9:00    -5.0   
Haze                                      3/13/2012 23:00    14.1   
Mainly Clear                                9/9/2012 9:00    33.0   
Moderate Rain,Fog                         12/10/2012 8:00     1.7   
Moderate Snow                             12/27/2012 9:00    -4.9   
Moderate Snow,Blowing Snow               12/27/2012 12:00    -5.4   
Mostly Cloudy                               9/9/2012 2:00    32.4   
Rain                                        9/5/2012 2:00    22.8   
Rain Showers                               9/8/2012 16:00    26.4   
Rain Showers,Fog                          10/20/2012 3:00    12.8   
Rain Showers,Snow Showers                 12/5/2012 10:00     2.2   
Rain,Fog                                  9/30/2012 23:00    21.7   
Rain,Haze                                  3/13/2012 9:00     5.5   
Rain,Ice Pellets                          12/18/2012 5:00     0.6   
Rain,Snow                                  4/23/2012 3:00     1.7   
Rain,Snow Grains                          12/21/2012 0:00     1.9   
Rain,Snow,Fog                             12/8/2012 21:00     0.8   
Rain,Snow,Ice Pellets                     12/21/2012 5:00     1.3   
Snow                                       4/27/2012 9:00     3.7   
Snow Pellets                             11/24/2012 15:00     0.7   
Snow Showers                               3/4/2012 21:00     2.9   
Snow Showers,Fog                         12/29/2012 13:00   -10.0   
Snow,Blowing Snow                          2/25/2012 9:00    -1.4   
Snow,Fog                                  3/14/2012 19:00     1.1   
Snow,Haze                                  2/1/2012 21:00    -3.6   
Snow,Ice Pellets                            3/3/2012 4:00     0.8   
Thunderstorms                              7/4/2012 16:00    26.7   
Thunderstorms,Heavy Rain Showers           5/29/2012 6:00    10.9   
Thunderstorms,Moderate Rain Showers,Fog    7/17/2012 6:00    19.6   
Thunderstorms,Rain                        7/23/2012 18:00    21.3   
Thunderstorms,Rain Showers                  9/8/2012 4:00    25.5   
Thunderstorms,Rain Showers,Fog            7/31/2012 20:00    22.9   
Thunderstorms,Rain,Fog                     7/17/2012 5:00    20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                      
Clear                                                20.4         99   
Cloudy                                               22.6         99   
Drizzle                                              17.7         96   
Drizzle,Fog                                          19.1        100   
Drizzle,Ice Pellets,Fog    

### Q.12 Show All the records where weather condition is Fog.


In [82]:
df[df["Weather Condition"]=='Fog']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
5        1/1/2012 5:00    -1.4              -3.3         87                9   
6        1/1/2012 6:00    -1.5              -3.1         89                7   
...                ...     ...               ...        ...              ...   
8716   12/29/2012 4:00   -16.0             -17.2         90                6   
8717   12/29/2012 5:00   -14.8             -15.9         91                4   
8718   12/29/2012 6:00   -13.8             -15.3         88                4   
8719   12/29/2012 7:00   -14.8             -16.4         88                7   
8722  12/29/2012 10:00   -12.0             -13.3         90                7   

      Visibility_km  Press_kPa Weather Condition  
0               8.0     101.24               Fog  
1               8.0     101.24               Fog  
4               4.8     101.23               Fog  
5               6.4     101.27               Fog  
6               6.4     101.29               Fog  
...             ...        ...               ...  
8716            9.7     101.25               Fog  
8717            6.4     101.25               Fog  
8718            9.7     101.25               Fog  
8719            8.0     101.22               Fog  
8722            6.4     101.15               Fog  

[150 rows x 8 columns]

### Q.13 Find all instances when 'Weather is clear' or 'Visibility is above 40'.

In [83]:
df[(df['Weather Condition']=='Clear') | (df['Visibility_km']>40)].head(50)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67    1/3/2012 19:00   -16.9             -24.8         50               24   
106   1/5/2012 10:00    -6.0             -10.0         73               17   
107   1/5/2012 11:00    -5.6             -10.2         70               22   
108   1/5/2012 12:00    -4.7              -9.6         69               20   
109   1/5/2012 13:00    -4.4              -9.7         66               26   
110   1/5/2012 14:00    -5.1             -10.7         65               22   
111   1/5/2012 15:00    -4.3             -12.0         55               26   
114   1/5/2012 18:00    -7.1             -14.4         56               11   
115   1/5/2012 19:00    -9.2             -15.4         61                7   
116   1/5/2012 20:00    -9.8             -15.7         62                9   
117   1/5/2012 21:00    -9.0             -14.8         63               13   
183   1/8/2012 15:00    -6.6             -12.9         61               20   
241   1/11/2012 1:00   -10.7             -17.8         56               17   
242   1/11/2012 2:00   -12.0             -18.9         56               19   
243   1/11/2012 3:00   -12.7             -19.4         57               19   
244   1/11/2012 4:00   -13.4             -20.1         57               17   
324  1/14/2012 12:00   -17.5             -23.8         58               20   
325  1/14/2012 13:00   -17.1             -24.1         55               17   
326  1/14/2012 14:00   -16.7             -23.4         56               17   
327  1/14/2012 15:00   -16.7             -23.4         56               22   
344   1/15/2012 8:00   -23.3             -28.5         62                7   
345   1/15/2012 9:00   -22.2             -27.8         60                9   
346  1/15/2012 10:00   -20.6             -26.8         58                9   
347  1/15/2012 11:00   -19.3             -26.1         55                9   
348  1/15/2012 12:00   -18.0             -25.5         52               13   
349  1/15/2012 13:00   -16.8             -24.2         53               20   
350  1/15/2012 14:00   -16.0             -23.4         53               26   
351  1/15/2012 15:00   -15.4             -22.8         53               24   
352  1/15/2012 16:00   -15.1             -22.8         52               24   
353  1/15/2012 17:00   -16.2             -23.2         55               15   
354  1/15/2012 18:00   -16.3             -22.9         57               17   
355  1/15/2012 19:00   -16.3             -22.7         58               20   
356  1/15/2012 20:00   -16.7             -22.4         61               17   
357  1/15/2012 21:00   -16.9             -22.2         63               15   
358  1/15/2012 22:00   -16.9             -21.7         66               17   
359  1/15/2012 23:00   -16.9             -21.4         68               15   
360   1/16/2012 0:00   -17.1             -21.9         66               15   
361   1/16/2012 1:00   -17.8             -22.7         65                6   
362   1/16/2012 2:00   -18.5             -23.0         68                6   
363   1/16/2012 3:00   -19.2             -23.8         67                6   
364   1/16/2012 4:00   -18.7             -23.5         66                0   
365   1/16/2012 5:00   -19.1             -23.3         69                0   
366   1/16/2012 6:00   -18.7             -23.2         68                0   
367   1/16/2012 7:00   -17.9             -21.5         73                4   
368   1/16/2012 8:00   -17.9             -22.6         67                9   
369   1/16/2012 9:00   -14.4             -20.7         59               22   
370  1/16/2012 10:00   -13.6             -18.7         65               19   
371  1/16/2012 11:00   -12.7             -17.7         66               20   
372  1/16/2012 12:00   -10.7             -16.9         60               19   
373  1/16/2012 13:00   -10.1             -16.1         61               13   

     Visibility_km  Press_kPa We

In [84]:
df[(df['Weather Condition']=='Clear') | (df['Visibility_km']>40)].tail(50)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
8387  12/15/2012 11:00    -9.3             -14.9         64               19   
8388  12/15/2012 12:00    -9.1             -15.1         62               19   
8389  12/15/2012 13:00    -8.4             -14.7         60               19   
8390  12/15/2012 14:00    -8.0             -14.2         61               13   
8391  12/15/2012 15:00    -7.8             -13.7         63               15   
8392  12/15/2012 16:00    -8.5             -14.8         60               20   
8394  12/15/2012 18:00    -9.1             -15.1         62               17   
8396  12/15/2012 20:00    -8.7             -15.1         60               20   
8408   12/16/2012 8:00    -9.5             -14.8         65               32   
8599   12/24/2012 7:00   -11.1             -13.9         80               15   
8600   12/24/2012 8:00   -11.0             -13.9         79               13   
8601   12/24/2012 9:00   -10.5             -13.7         77               13   
8602  12/24/2012 10:00    -9.9             -13.4         76               11   
8603  12/24/2012 11:00    -9.0             -13.7         69               11   
8604  12/24/2012 12:00    -7.9             -13.3         65                9   
8605  12/24/2012 13:00    -7.6             -13.1         65               15   
8606  12/24/2012 14:00    -7.8             -13.7         63               15   
8607  12/24/2012 15:00    -7.5             -13.3         63               13   
8610  12/24/2012 18:00   -10.4             -13.8         76                9   
8630  12/25/2012 14:00    -7.7             -14.1         60                6   
8631  12/25/2012 15:00    -7.1             -13.7         59               17   
8632  12/25/2012 16:00    -7.5             -13.9         60               11   
8633  12/25/2012 17:00    -8.3             -13.4         67               13   
8637  12/25/2012 21:00    -9.7             -12.5         80                4   
8638  12/25/2012 22:00   -10.9             -13.2         83                4   
8639  12/25/2012 23:00   -10.4             -12.7         83               11   
8640   12/26/2012 0:00   -11.8             -13.5         87                4   
8641   12/26/2012 1:00   -11.2             -12.9         87                6   
8642   12/26/2012 2:00   -12.7             -14.4         87                4   
8643   12/26/2012 3:00   -14.2             -15.8         88                6   
8644   12/26/2012 4:00   -13.1             -14.7         88                6   
8645   12/26/2012 5:00   -12.7             -14.1         89                4   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8651  12/26/2012 11:00   -11.3             -14.5         77               20   
8652  12/26/2012 12:00   -10.6             -14.3         74               20   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8699  12/28/2012 11:00    -6.2              -8.8         82               24   
8700  12/28/2012 12:00    -7.2              -9.9         81               24   
8701  12/28/2012 13:00    -6.8              -9.8         79               20   
8702  12/28/2012 14:00    -6.5              -9.9         77               22   
8703  12/28/2012 15:00    -6.8             -10.3         76               24   
8704  12/28/2012 16:00    -7.7             -11.0         77               30   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8748  12/30/2012 12:00   -12.2             -15.7         75               26   
8749  12/30/2012 13:00   -12.4             -16.2         73               37   
8750  12/30/2012 14:00   -11.8             -16.1         70               37   
8751  12/30/2012 15:00   -11.3             -15.6         70               32   
8752  12/30/2012 16:00   -11.4             -15.5         72               26   
8756  12/30/2

### Q.14 Find all instances when:
a) 'Weather is clear' and 'Relative humidity is greater than 50' or

b) 'Visibility' is above 40

In [85]:
df[(df['Weather Condition']=='Clear' ) & (df['Rel Hum_%'] >50) | (df['Visibility_km']> 40)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
106     1/5/2012 10:00    -6.0             -10.0         73               17   
107     1/5/2012 11:00    -5.6             -10.2         70               22   
108     1/5/2012 12:00    -4.7              -9.6         69               20   
109     1/5/2012 13:00    -4.4              -9.7         66               26   
110     1/5/2012 14:00    -5.1             -10.7         65               22   
...                ...     ...               ...        ...              ...   
8749  12/30/2012 13:00   -12.4             -16.2         73               37   
8750  12/30/2012 14:00   -11.8             -16.1         70               37   
8751  12/30/2012 15:00   -11.3             -15.6         70               32   
8752  12/30/2012 16:00   -11.4             -15.5         72               26   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather Condition  
106            48.3     100.45      Mainly Clear  
107            48.3     100.41      Mainly Clear  
108            48.3     100.38      Mainly Clear  
109            48.3     100.40      Mainly Clear  
110            48.3     100.46      Mainly Clear  
...             ...        ...               ...  
8749           48.3     100.92     Mostly Cloudy  
8750           48.3     100.96      Mainly Clear  
8751           48.3     101.05      Mainly Clear  
8752           48.3     101.15      Mainly Clear  
8756           25.0     101.52             Clear  

[2921 rows x 8 columns]